### PR TITLE
[SYCL][Graph] Avoid unnecessary inter-partition dependencies

### PR DESCRIPTION
**Pull request by mfrancepillois**
Improves management of inter-partition dependencies, so that only required dependencies are added. 
As removing these dependencies can results in multiple executions paths, we have added a map to track all events returned from submitted partitions. 
All these events are linked to the main event returned to user. 
Adds tests.

### CODE REVIEW

File: sycl/source/detail/graph_impl.hpp [hunk 1]
```diff
@@ -496,6 +496,7 @@ class partition {
   /// Map of devices to command buffers.
   std::unordered_map<sycl::device, sycl::detail::pi::PiExtCommandBuffer>
       MPiCommandBuffers;
+  std::vector<std::shared_ptr<partition>> MPredecessors;
```
Comment (EwanC) [inline sycl/source/detail/graph_impl.hpp:500]: doxy comment

File: sycl/test-e2e/Graph/RecordReplay/host_task2_multiple_roots.cpp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{env UR_L0_LEAKS_DEBUG=1 %{run} %t.out 2>&1 | FileCheck %s %}
+//
+// CHECK-NOT: LEAK
```
Comment (EwanC) [inline sycl/test-e2e/Graph/RecordReplay/host_task2_multiple_roots.cpp]: prefer `--implicit-check-not=LEAK`

**Review by EwanC — APPROVED** [review]
This change looks reasonable, main comment I have is that it would be easier to review and maintain if there was a diagram somewhere of before and after this change.

File: sycl/doc/design/CommandGraph.md [hunk 3]
```diff
@@ -192,6 +192,32 @@ illustrated in the following diagrams:
 ![Graph partition illustration step 10b.](images/SYCL-Graph-partitions_step11.jpg)
 ![Graph partition illustration step 11b.](images/SYCL-Graph-partitions_step12.jpg)
 
+### Multiple roots execution flow
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]: Nitpick, but think other subheadings are stylized like:
```suggestion
### Multiple Roots Execution Flow
```

File: sycl/doc/design/CommandGraph.md [hunk 4]
```diff
@@ -192,6 +192,32 @@ illustrated in the following diagrams:
...
+both branches can be executed concurrently.
+But as we can see on this diagram, this new approach can involve 
+multiple execution tails, which leads to difficulties when 
+we want to know when the graph execution has finished.
+To cope with this issue, the events associated to the completion of 
+each partition are linked to the event returned to users. 
+Hence, when the returned event is complete, we can guarantee that 
+all works associated with the graph has been completed.  
+
+[Multiple roots graph partition execution flow.](images/SYCL-Graph-partition_execution_flow.jpg)
```
Comment (EwanC) [inline sycl/doc/design/CommandGraph.md]: This diagram isn't rendering, think it needs to be
```suggestion
![Multiple roots graph partition execution flow.](images/SYCL-Graph-partition_execution_flow.jpg)
```

**Review by EwanC — APPROVED** [review]
Documentation is a great addition 👍

**Review by Bensuo — APPROVED** [review]

**Comment by mfrancepillois** [comment]
Upstream PR: https://github.com/intel/llvm/pull/12680

**Pull request closed without merging**